File: Camp Progress/200. Number of Islands(with Union Find).cpp
// Question Link: https://leetcode.com/problems/number-of-islands/

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        
        unordered_map<string, string> dsu;
        
        int M = grid.size();
        int N = grid[0].size();
        
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                if (grid[i][j] == '1') {
                    string a = to_string(i) + "," + to_string(j);
                    dsu[a] = a;
                    // Check left
                    if (j - 1 >= 0 && grid[i][j - 1] == '1') {
                        string b = to_string(i) + "," + to_string(j - 1);
                        Union(a, b, dsu);
                    }
                    // Check Up
                    if (i - 1 >= 0 && grid[i - 1][j] == '1') {
                        string b = to_string(i - 1) + "," + to_string(j);
                        Union(a, b, dsu);
                    }
                    
                }
            }   
        }
        
        int numOfIslands = 0;
        for (auto &it : dsu) {
            if (it.second == it.first) {
                numOfIslands++;
            }
        }
        
        return numOfIslands;
    }
    
    string find(string cur, unordered_map<string, string> &dsu) {
        if (dsu[cur] == cur) {
            return cur;
        }
        return find(dsu[cur], dsu);
    }    
    
    void Union(string &a, string &b, unordered_map<string, string> &dsu) {
        string parent = find(a, dsu);
        string child = find(b, dsu);
        if (parent != child) {
            dsu[parent] = child;            
        }
    }
};

File: Python3 Solutions/200. Number of Islands(with Union Find).py
# Question Link: https://leetcode.com/problems/number-of-islands/

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        
        dsu = {}
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    key = str(i) + ',' + str(j)
                    dsu[key] = key
                    if j - 1 >= 0 and grid[i][j - 1] == '1':
                        b = str(i) + ',' + str(j - 1)
                        self.union(key, b, dsu)
                    if i - 1 >= 0 and grid[i - 1][j] == '1':
                        b = str(i - 1) + ',' + str(j)
                        self.union(key, b, dsu)
        
        num_of_islands = 0
        for key, val in dsu.items():
            if key == val:
                num_of_islands += 1
        
        return num_of_islands
    
    def find(self, cur, dsu):
        if dsu[cur] == cur:
            return cur
        return self.find(dsu[cur], dsu)
    
    def union(self, a, b, dsu):
        parent_a = self.find(a, dsu)
        parent_b = self.find(b, dsu)
        dsu[parent_a] = parent_b
